src/cart/src/cart/crt_rpc.c

```diff
@@ -536,251 +536,259 @@ crt_req_fill_tgt_uri(struct crt_rpc_priv *rpc_priv, crt_phy_addr_t base_uri)
-{
-	struct crt_context	*crt_ctx;
-	crt_rpc_t		*rpc_pub = &rpc_priv->crp_pub;
-	int			 rc;
-
-	/* destroy previous URI_LOOKUP req, addref in crt_req_uri_lookup_psr */
-	RPC_PUB_DECREF(rpc_priv->crp_ul_req);
-	rpc_priv->crp_ul_req = NULL;
-
-	rc = crt_grp_psr_reload(grp_priv);
-	if (rc != 0) {
-		RPC_ERROR(rpc_priv,
-			  "crt_grp_psr_reload(grp %s) failed, rc: %d\n",
-			  grp_priv->gp_pub.cg_grpid, rc);
-		D_GOTO(out, rc);
-	}
-
-	crt_ctx = rpc_pub->cr_ctx;
-
-	D_MUTEX_LOCK(&crt_ctx->cc_mutex);
-	if (!crt_req_timedout(rpc_priv))
-		crt_req_timeout_untrack(rpc_priv);
-	crt_set_timeout(rpc_priv);
-	rc = crt_req_timeout_track(rpc_priv);
-	D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
-	if (rc != 0) {
-		RPC_ERROR(rpc_priv,
-			  "crt_req_timeout_track failed, rc: %d\n",
-			  rc);
-		D_GOTO(out, rc);
-	}
-
-	rpc_priv->crp_state = RPC_STATE_INITED;
-	rc = crt_req_send_internal(rpc_priv);
-
-out:
-	return rc;
-}
+static int
+crt_issue_uri_lookup(crt_context_t ctx, crt_group_t *group,
+		     d_rank_t contact_rank, uint32_t contact_tag,
+		     d_rank_t query_rank, uint32_t query_tag,
+		     struct crt_rpc_priv *chained_rpc);
 
 static void
-crt_req_uri_lookup_by_rpc_cb(const struct crt_cb_info *cb_info)
+uri_lookup_cb(const struct crt_cb_info *cb_info)
 {
-	d_rank_t			 rank;
-	uint32_t			 tag;
-	crt_endpoint_t			*tgt_ep;
-	struct crt_rpc_priv		*rpc_priv;
-	struct crt_grp_priv		*grp_priv;
-	struct crt_context		*crt_ctx;
+	struct crt_rpc_priv		*chained_rpc_priv;
+	struct crt_context		*ctx;
 	struct crt_uri_lookup_out	*ul_out;
-	char				*uri = NULL;
-	bool				 ul_retried = false;
-	int				 rc = 0;
-
-	rpc_priv = cb_info->cci_arg;
-	D_ASSERT(rpc_priv->crp_state == RPC_STATE_URI_LOOKUP ||
-		 rpc_priv->crp_state == RPC_STATE_TIMEOUT ||
-		 rpc_priv->crp_state == RPC_STATE_FWD_UNREACH ||
-		 rpc_priv->crp_state == RPC_STATE_COMPLETED);
-
-	/* Failed uri lookup can complete with cb_info->rpc pointing to
-	 * the original RPC that triggered the lookup if HG_Forward
-	 * fails with mercury errors.
-	 */
-	if (cb_info->cci_rc == DER_SUCCESS)
-		D_ASSERT(rpc_priv->crp_ul_req == cb_info->cci_rpc);
-
-	tgt_ep = &rpc_priv->crp_pub.cr_ep;
-	rank = tgt_ep->ep_rank;
-	tag = tgt_ep->ep_tag;
-
-	grp_priv = crt_grp_pub2priv(tgt_ep->ep_grp);
+	struct crt_uri_lookup_in	*ul_in;
+	struct crt_grp_priv		*grp_priv;
+	crt_rpc_t			*lookup_rpc;
+	int				rc = 0;
 
-	crt_ctx = rpc_priv->crp_pub.cr_ctx;
+	chained_rpc_priv = cb_info->cci_arg;
+	lookup_rpc  = cb_info->cci_rpc;
 
 	if (cb_info->cci_rc != 0) {
-		RPC_ERROR(rpc_priv,
-			  "failed cci_rc: %d\n",
-			  cb_info->cci_rc);
-		if (cb_info->cci_rc == -DER_OOG)
-			D_GOTO(out, rc = -DER_OOG);
-
-		/* TODO: Proper selection of new PSR is required */
-		rc = cb_info->cci_rc;
-		D_GOTO(out, rc);
+		D_ERROR("URI_LOOKUP rpc completed with rc=%d\n",
+			cb_info->cci_rc);
+		D_GOTO(retry, rc = cb_info->cci_rc);
 	}
 
-	/* extract uri */
-	ul_out = crt_reply_get(cb_info->cci_rpc);
-	D_ASSERT(ul_out != NULL);
+	ul_in = crt_req_get(lookup_rpc);
```

Comment:
we generally prefer lookup_rpc->cr_input although in practice it doesn't make much difference.